### PR TITLE
Workarounds for deadlocks (and follow-on issues) observed during channel close

### PR DESCRIPTION
**Pull request by mpontillo**
This PR attempts to work around three issues:

* Calling `c.Close()` within the `OnDisconnection` handler is redundant, and leads to a recursive call and a deadlock.
* Calling `close(ch)` upon receiving an `atlas_error` is redundant because the channel will be closed within the `OnDisconnection` handler as well.
* It is possible, given the current overall design of the code, that a write to a closed channel could be attempted. Use `recover()` to prevent a crash. (A larger effort would be needed to solve this in a "more correct" way.)

### CODE REVIEW

**Comment by mpontillo** [comment]
Apologies if this PR is a bit premature; I intended to PR it internally in the `digitalocean` fork.
Also, I can't take credit for all these changes; I paired with one of my colleagues to come up with these workarounds. Our goal is to solve this issue:
https://github.com/czerwonk/atlas_exporter/issues/12

**Comment by mpontillo** [comment]
Here's our internal PR, for the record. We're still reviewing and testing this internally.
https://github.com/digitalocean/ripeatlas/pull/1

**Comment by jelu** [comment]
Cool, didn't know this was used out there like this.

Let me know how the findings goes. I won't be able to look at this for some weeks anyway.

File: stream.go [hunk 1]
```diff
@@ -147,7 +145,7 @@ func (s *Stream) MeasurementLatest(p Params) (<-chan *measurement.Result, error)
         err := h.Emit("atlas_subscribe", subscribe)
         if err != nil {
             r := &measurement.Result{ParseError: fmt.Errorf("h.Emit(atlas_subscribe): %s", err.Error())}
-            ch <- r
+            trySend(ch, r)
             c.Close()
             close(ch)
```
Comment (glightfoot) [inline stream.go:150]: I think this close is redundant, since the channel is closed by the onDisconnection method called by c.Close()

Comment (mpontillo) [reply stream.go:150]: Nice catch Greg; let's land this change on our internal fork so we can test with it.

**Comment by mpontillo** [comment]
Thanks @jelu; we're happy to help.
This branch is now obsolete now that we've merged it into our internal fork, so I've created a new draft PR for your eventual consideration (#22), and will go ahead and close this one.

**Pull request closed without merging**